Bottom: cfae713d7af004862b8134efb3e08a66e14b8da5
Top:    b58d3427f335e43edf6246d4a05028b8a4e791de
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-05-13 14:31:34 +0100

hid/common/hidgl.c: Use tags on circular contours to render them faster

This lets us avoid the polygon tesselator for circular contours.


---

diff --git a/src/hid/common/hidgl.c b/src/hid/common/hidgl.c
index 00efdce..bbba217 100644
--- a/src/hid/common/hidgl.c
+++ b/src/hid/common/hidgl.c
@@ -554,11 +554,24 @@ hidgl_fill_polygon (int n_coords, int *x, int *y)
 }
 
 void
-tesselate_contour (GLUtesselator *tobj, PLINE *contour, GLdouble *vertices)
+tesselate_contour (GLUtesselator *tobj, PLINE *contour, GLdouble *vertices,
+                   double scale)
 {
   VNODE *vn = &contour->head;
   int offset = 0;
 
+  /* If the contour is round, and hidgl_fill_circle would use
+   * less slices than we have vertices to draw it, then call
+   * hidgl_fill_circle to draw this contour.
+   */
+  if (contour->is_round) {
+    double slices = calc_slices (contour->radius / scale, 2 * M_PI);
+    if (slices < contour->Count) {
+      hidgl_fill_circle (contour->cx, contour->cy, contour->radius, scale);
+      return;
+    }
+  }
+
   gluTessBeginPolygon (tobj, NULL);
   gluTessBeginContour (tobj);
   do {
@@ -575,6 +588,7 @@ tesselate_contour (GLUtesselator *tobj, PLINE *contour, GLdouble *vertices)
 struct do_hole_info {
   GLUtesselator *tobj;
   GLdouble *vertices;
+  double scale;
 };
 
 static int
@@ -588,7 +602,7 @@ do_hole (const BoxType *b, void *cl)
     return 0;
   }
 
-  tesselate_contour (info->tobj, curc, info->vertices);
+  tesselate_contour (info->tobj, curc, info->vertices, info->scale);
   return 1;
 }
 
@@ -605,6 +619,7 @@ hidgl_fill_pcb_polygon (PolygonType *poly, const BoxType *clip_box, double scale
   struct do_hole_info info;
   int stencil_bit;
 
+  info.scale = scale;
   global_scale = scale;
 
   if (poly->Clipped == NULL)
@@ -664,7 +679,7 @@ hidgl_fill_pcb_polygon (PolygonType *poly, const BoxType *clip_box, double scale
   /* Drawing operations as masked to areas where the stencil buffer is '0' */
 
   /* Draw the polygon outer */
-  tesselate_contour (info.tobj, poly->Clipped->contours, info.vertices);
+  tesselate_contour (info.tobj, poly->Clipped->contours, info.vertices, scale);
   hidgl_flush_triangles (&buffer);
 
   /* Unassign our stencil buffer bit */
